### DISCUSSION TITLE
Project Update: Open Cap Stack - Lake House

### DISCUSSION THREAD

**Discussion by urbantech** [Open-Cap-Stack/opencap#96] [General]
### **Overview**
In this latest sprint, additional significant progress was made in the integration and testing of the Lake House architecture for the Open Cap Stack application. The key components—PostgreSQL, MinIO (for object storage), and Airflow (for workflow orchestration)—have been successfully integrated. A comprehensive suite of tests was conducted to validate each component's functionality. Below is a detailed breakdown of all tasks completed, additional work carried out, and outcomes achieved.

---

### **1. Airflow Integration**
- **Setup**: Airflow has been configured to orchestrate DAGs (Directed Acyclic Graphs) for automating data processing workflows.
- **DAG Creation**: A test DAG (`test_dag`) was created, added to the Airflow UI, and validated to ensure it was recognized in the system.
- **API Testing**: I triggered the DAG using Airflow’s REST API to confirm that the system could queue and process the DAG.

  **New Test**:
  - A new DAG was successfully triggered via the Airflow API.
  - The API returned a `200 OK` status, with the DAG execution state queued and processed as expected.

  **Outcome**: The Airflow integration, along with API-triggered DAG execution, has been confirmed to be working as expected.

---

### **2. Data Processing Pipeline**
- **Objective**: Implement and validate the data processing pipeline, which reads and processes a test dataset.
- **Approach**:
  - A test dataset (`test-dataset.csv`) was utilized to simulate a real-world dataset.
  - Custom processing logic was applied to the dataset to validate data transformations.
  
  **New Work**:
  - Data processing was further refined with more complex transformation logic.
  - The test successfully produced expected data outputs, confirming the integrity of the pipeline.

  **Outcome**: The data processing pipeline processed the dataset correctly, with the expected transformations validated.

---

### **3. Metadata Logging in PostgreSQL**
- **Setup**: PostgreSQL serves as the metadata store for logging dataset information processed by the system.
- **Schema**: The schema includes fields for dataset name, description, storage location, creation time, and last modified time.
- **New Tests**:
  - New datasets were logged to PostgreSQL, and the metadata was cross-checked for accuracy.
  - The database schema was expanded to capture additional metadata related to dataset processing.

  **Outcome**: PostgreSQL successfully logged the dataset metadata, confirming that the schema and logging mechanisms are functioning correctly.

---

### **4. Object Storage with MinIO**
- **Objective**: Store datasets and files in MinIO, a local object storage service.
- **Integration**: MinIO was integrated as the primary storage backend.
- **Storage Workflow**:
  - Datasets were uploaded to the `lakehouse-bucket` in MinIO.
  - Storage locations were logged in PostgreSQL to maintain data lineage and traceability.

  **New Work**:
  - Environment variables for MinIO were configured (`MINIO_ENDPOINT`, `MINIO_ACCESS_KEY`, and `MINIO_SECRET_KEY`).
  - A file upload test was executed, verifying that MinIO stored the dataset and logged its metadata.

  **Outcome**: The MinIO integration successfully stores datasets, and metadata is accurately captured in PostgreSQL.

---

### **5. Additional Setup & Environment Configuration**
- **Environment Configuration**: A `.env` file was configured with MinIO credentials, and the environment was validated to ensure that MinIO keys were accessible to the application.
- **MinIO Admin Setup**: Admin credentials were successfully configured to interact with MinIO, and an alias was set up using the MinIO CLI for easier bucket management.

  **Outcome**: MinIO environment setup was confirmed, allowing seamless file storage and bucket management.

---

### **6. Testing & Validation**
I ran several integration tests to validate the functionality of the system end-to-end:

1. **Airflow Integration Test**:
   - **Test File**: `airflowIntegration.test.js`
   - **Status**: Passed (DAG was successfully triggered via the API).

2. **MinIO Integration Test**:
   - **Test File**: `minioIntegration.test.js`
   - **Status**: Passed (Dataset was logged to PostgreSQL and stored in MinIO).
  
3. **PostgreSQL Logging Test**:
   - **Test File**: `postgresIntegration.test.js`
   - **Status**: Validated that datasets and metadata are being logged correctly in PostgreSQL.

---

### **Next Steps**
1. **Data Storage Automation**: Automate file uploads to MinIO and enhance the processing pipeline to handle production-scale datasets.
2. **Advanced Metadata Logging**: Expand the PostgreSQL schema to include more detailed metadata fields, such as processing times, data lineage, and user interactions.
3. **Larger Dataset Processing**: Test the system's performance with larger datasets to ensure stability and scalability.
4. **Additional Workflow Orchestration**: Integrate new DAGs and workflows for more complex data processing tasks.

---

### **Conclusion**
The Open Cap Stack Lake House architecture is now fully operational, with seamless integration between Airflow, PostgreSQL, and MinIO. All key components have been validated through extensive testing, and the system is ready to handle larger datasets and more complex workflows in future iterations.

---

Here’s a detailed, itemized list of all the libraries, frameworks, and tools that have been installed to date in order to set up the Lake House demo for the Open Cap Stack application. This list includes environment configurations, services, and key packages for local setup.

---

## **1. System Environment & Tools**
- **Python (v3.8.10)**:
  - Installed via Pyenv for managing Python versions.
  - Ensure to install Pyenv and set Python 3.8.10 as the active version:
    ```bash
    pyenv install 3.8.10
    pyenv global 3.8.10
    ```
  
- **Node.js (v14.x or later)**:
  - Installed using NVM (Node Version Manager) or directly from Node.js downloads.

- **MinIO Server**:
  - A self-hosted, S3-compatible object storage service for storing datasets.
  - Install MinIO via the official site, and run the MinIO server locally with:
    ```bash
    minio server /path/to/minio-data --console-address :9001
    ```

- **MinIO Client (mc)**:
  - CLI tool to manage MinIO and other object storage systems.
  - Install via:
    ```bash
    brew install minio/stable/mc
    ```

- **Airflow (v2.x)**:
  - Airflow is used for orchestrating workflows.
  - Installed using pip:
    ```bash
    pip install apache-airflow
    ```

## **2. Python Packages & Libraries**
To ensure all necessary Python libraries are installed, you can create a `requirements.txt` file with the following dependencies:

- **Airflow** (for orchestration):
  ```bash
  pip install apache-airflow
  ```

- **MinIO Python SDK** (for object storage interaction):
  ```bash
  pip install minio
  ```

- **psycopg2-binary** (for PostgreSQL interaction):
  ```bash
  pip install psycopg2-binary
  ```

- **dotenv** (for environment variable management):
  ```bash
  pip install python-dotenv
  ```

- **Flask** (optional if you want to run API services):
  ```bash
  pip install flask
  ```

- **Pytest / Jest** (for running tests):
  - Jest is used for Node.js-based integration tests:
    ```bash
    npm install --save-dev jest
    ```

    For Python tests, if necessary:
    ```bash
    pip install pytest
    ```

## **3. Node.js & JavaScript Packages**
- **axios** (for making HTTP requests in tests and API calls):
  ```bash
  npm install axios
  ```

- **dotenv** (for environment variable management in Node.js):
  ```bash
  npm install dotenv
  ```

- **Jest** (for testing):
  ```bash
  npm install --save-dev jest
  ```

## **4. PostgreSQL Setup**
- **PostgreSQL** (for metadata storage):
  - Install PostgreSQL via Homebrew (or use Docker):
    ```bash
    brew install postgresql
    ```

- **PgAdmin** (for PostgreSQL management):
  - Optionally, install PgAdmin to manage PostgreSQL databases via a GUI.
  - Installed via Homebrew:
    ```bash
    brew install --cask pgadmin4
    ```

## **5. Environment Variables Configuration**
You need an `.env` file in your project root that contains the following key environment variables for MinIO and PostgreSQL:

```bash
# .env file example
# MinIO Configurations
MINIO_ENDPOINT=localhost:9000
MINIO_ACCESS_KEY=minioadmin
MINIO_SECRET_KEY=minioadmin

# PostgreSQL Configurations
POSTGRES_DB=open_cap_stack
POSTGRES_USER=postgres
POSTGRES_PASSWORD=yourpassword
POSTGRES_HOST=localhost
POSTGRES_PORT=5432

# Airflow Admin
AIRFLOW_USERNAME=admin
AIRFLOW_PASSWORD=admin_password
```

Make sure you source this `.env` file in your local environment before running the system:
```bash
source /path/to/your/.env
```

## **6. Workflow to Start Services**
1. **Start PostgreSQL**:
   - Ensure PostgreSQL is running:
   ```bash
   brew services start postgresql
   ```

2. **Start MinIO**:
   - Start the MinIO server:
   ```bash
   minio server /path/to/minio-data --console-address :9001
   ```

3. **Start Airflow**:
   - Start Airflow’s webserver:
   ```bash
   airflow webserver --port 8081
   ```
   - Start Airflow’s scheduler:
   ```bash
   airflow scheduler
   ```

## **7. Tests and Validation**
The following test files have been set up to ensure proper integration of the Lake House components:

- **Airflow Integration Test**: `airflowIntegration.test.js`
- **MinIO Integration Test**: `minioIntegration.test.js`
- **PostgreSQL Logging Test**: `postgresIntegration.test.js`

Run the tests using:
```bash
npx jest __tests__/airflowIntegration.test.js
npx jest __tests__/minioIntegration.test.js
npx jest __tests__/postgresIntegration.test.js
```

---

### **Other Tools to consider**

For the "Lake House" project, several tools, libraries, and frameworks can complement your existing architecture with **GraphDB** and **Apache Spark**. Here’s a curated list categorized by functionality:

### **1. Data Ingestion & ETL:**
- **Apache NiFi** or **Apache Kafka**: For real-time data ingestion, streaming, and ETL (Extract, Transform, Load) capabilities. These tools can automate and streamline the process of ingesting structured and unstructured data into your lake house.
- **dbt (Data Build Tool)**: Ideal for transforming data in the lake house. It works well with Spark and can help manage and automate data transformations.

### **2. Data Processing & Storage:**
- **Delta Lake**: Already part of your setup, but critical to highlight. It adds **ACID** transactions to Spark, making data storage reliable and consistent, supporting schema evolution and version control.
- **Hudi** or **Iceberg**: Alternative options to Delta Lake if you need more flexibility or specific integration features with other systems.

### **3. Data Governance & Metadata Management:**
- **Apache Atlas**: For metadata management and governance. It integrates well with Apache Spark, providing data lineage and metadata tracking, which is crucial for regulated industries and auditing purposes.
- **Amundsen**: An open-source data discovery and metadata engine developed by Lyft, which provides easy navigation and discovery of your datasets within the lake house.

### **4. Knowledge Graph & Semantic Layer:**
- **GraphDB Plugins**: Explore additional plugins that can enhance GraphDB’s inference and reasoning capabilities. Some of these plugins can provide ontology-driven data validation and semantic enrichment.
- **Apache Jena**: If you need additional flexibility or alternatives for RDF storage, reasoning, or SPARQL execution alongside GraphDB.

### **5. Orchestration & Workflow Automation:**
- **Apache Airflow**: You already have this in place, but you can expand its capabilities with additional DAGs to automate data ingestion, processing, and analytics tasks.
- **Prefect**: A modern alternative to Airflow with a more Pythonic approach, making it easier to define and maintain workflows.

### **6. Data Visualization & Exploration:**
- **Apache Superset**: An open-source data exploration and visualization platform. It integrates well with various data sources, including Spark, PostgreSQL, and GraphDB, and provides a UI for visualizing insights.
- **Jupyter Notebooks**: Essential for ad-hoc analysis and exploring the knowledge graph and processed data. Integration with Spark and GraphDB is straightforward.

### **7. Machine Learning & Advanced Analytics:**
- **MLflow**: A tracking tool to manage ML experiments and deployments. It works well with Apache Spark and allows you to monitor and version control ML models.
- **TensorFlow Extended (TFX)** or **PyTorch Lightning**: If machine learning pipelines are a key component, these frameworks can help build robust and scalable ML pipelines on top of Spark and the lake house.

### **8. Security & Access Control:**
- **Apache Ranger**: Provides centralized security administration for managing access controls and policies across various services like Spark, PostgreSQL, MinIO, and GraphDB.
- **OAuth2 and OpenID Connect (OIDC)**: Implement these for securing REST APIs and web-based interfaces for accessing and managing data in your lake house.

### **9. Data Search & Indexing:**
- **Elasticsearch**: For real-time search and analytics on unstructured or semi-structured data. It can integrate with Spark and complement GraphDB for full-text search capabilities.
- **Apache Solr**: An alternative to Elasticsearch, with strong indexing and query capabilities for large-scale data collections.

### **10. Monitoring & Observability:**
- **Prometheus** & **Grafana**: For monitoring and visualizing metrics of your lake house components like Spark jobs, MinIO, PostgreSQL, and Airflow workflows.
- **ELK Stack (Elasticsearch, Logstash, Kibana)**: For centralized logging and monitoring of data ingestion, processing pipelines, and application health.

### Summary
Considering these tools and libraries can help you achieve a comprehensive "Lake House" architecture with robust ingestion, processing, governance, visualization, machine learning, security, and monitoring capabilities. This list aligns well with your project's data-centric goals and integrates seamlessly into the current stack.
